dm/relay/meta.go

```diff
@@ -212,7 +212,7 @@ func (lm *LocalMeta) Save(pos mysql.Position, gset gtid.Set) error {
 		lm.BinlogGTID = ""
 	} else {
 		lm.BinlogGTID = gset.String()
-		lm.gset = gset
+		lm.gset = gset.Clone() // need to clone and set, in order to avoid the local meta's gset and the input gset referencing the same object, causing contentions later
```

Comment:
maybe just add a lock in `func (r *Relay) SaveMeta(pos mysql.Position, gset gtid.Set)` is more intuitive 😬

Reply:
Adding a lock on `(*Relay).SaveMeta()` won't work.  Because the data race happens when the doing the `(gtid.Set).Set()` and `(gtid.Set).Clone()` on the same object.  Even if `(*Relay).SaveMeta()` is executed serially, data race will still happen when the input `gtid.Set` object is also referenced by the `LocalMeta` afterwards.  

Actually, there are two solutions on the problem.  The first one is to add a locking mechanism on corresponding methods in `gtid.Set` implementation.  The second one is to ensure that the `gtid.Set` object inside `LocalMeta` is independent.  

The first method involves many modifications and may have some performance penalty.  What's more, some current mechanism in `LocalMeta` tends to favor the second method.  For example, `(*LocalMeta).GTID()` clones the `gtid.Set` before return, which intends to isolate the returning `gtid.Set` object from the object inside `LocalMeta`. 

So considering only the `gtid.Set` object  inside `LocalMeta` is accessed frequently, and currently there are no places where the `gtid.Set` inside `LocalMeta` is updated directly (which means `(*LocalMeta).gset` is always updated by assigning a new object, rather than calling `(*LocalMeta).gset.Set()` ), I used the second method to fix this problem.

Reply:
thanks explain this reason i will read more carefully later

Reply:
this commets [help](https://github.com/pingcap/tiflow/issues/4166#issuecomment-1018476045), i got the reason of datarace now ~